### PR TITLE
[backend] feature(freetrials/chunk3): create trial request #1095

### PR DESCRIPTION
**Pull request by carinelebas**
# Context
> Briefly describe the purpose and context of this PR. Include relevant screenshots or screen recordings for the product team.

We need to have a graphql endpoint that will create a freetrial request.

on Apollo sandbox, call the following mutation and check the result. 
<details>

```

mutation CreateDeploymentRequest {
  createDeploymentRequest(
    input: {
      platform_identifier: opencti
      region: europe
      type: trial
      intention: "Evaluating OpenCTI for threat intelligence management in our security operations"
      job_title: "Security Analyst"
      activity_sector: "Financial Services"
      use_case: "use case"
    }
  ) {
    id
    platform_identifier
    region
    type
    intention
    job_title
    activity_sector
    start_date
    end_date
    status
  }
}

```

</details>

## Related issues

- Related to #1095

# Checklist

## Quality

- [x] I consider this work complete and ready for review
- [x] I tested all features and edge cases
- [x] I refactored code where necessary to improve quality
- [x] I made sure my code meets development guidelines
- [x] I performed a self-review of my code

## Testing

- [x] I wrote test cases for the relevant use cases (unit, integration or e2e tests)
- [ ] (Bug fixes only) I added tests that reproduce and verify the fix

### What tests have been made

- [ ] Unit tests
- [x] Integration tests
- [ ] E2E tests
- [x] Local manual tests

# Additional information
> Any additional context, dependencies, or concerns reviewers should know about.

### CODE REVIEW

File: apps/portal-api/src/modules/services/deployments/deployments.app.ts [hunk 1]
```diff
@@ -1,24 +1,77 @@
...
+      const [createdDeploymentRequest] = await insertDeploymentRequest({
+        id: uuidv4() as DeploymentRequestId,
+        user_requester_id: context.user.id,
+        organization_requester_id: context.user.selected_organization_id,
+        service_instance_id: serviceInstanceId,
+        status: DeploymentRequestStatus.Pending,
+        type: input.type,
+        platform_identifier: input.platform_identifier,
+        region: input.region,
+        intention: input.intention,
```
Comment (hervyt) [inline apps/portal-api/src/modules/services/deployments/deployments.app.ts]: **thought**: idk what this is...

File: apps/portal-api/src/modules/services/deployments/deployments.app.ts [hunk 2]
```diff
@@ -1,24 +1,77 @@
...
+        job_title: input.job_title,
+        use_case: input.use_case,
+        activity_sector: input.activity_sector,
+        platform_token: uuidv4(),
+      });
+      await trx.commit();
+      return {
+        id: createdDeploymentRequest.id,
```
Comment (hervyt) [inline apps/portal-api/src/modules/services/deployments/deployments.app.ts]: **question**: Qhy can't you directly return createdDeploymentRequest and have to rebuild the object ?

File: apps/portal-api/knexfile.ts [hunk 3]
```diff
@@ -148,8 +150,9 @@ export const db = <T>(
     opts
   );
 
-  if (context?.trx) {
-    queryContext.transacting(context.trx);
+  const reqContext = requestContext.get();
+  if (reqContext?.trx && !reqContext.trx.isCompleted()) {
+    queryContext.transacting(reqContext.trx);
```
Comment (Kroustille) [inline apps/portal-api/knexfile.ts]: Maybe we can add an helper function which does this because I think we'll need it in a lot of different places, and I'm afraid we're trying to use `transaction` on a completed transaction sometimes 😬

File: apps/portal-api/src/modules/services/instances/domain.ts [hunk 4]
```diff
@@ -12,15 +15,16 @@ export const serviceInstanceDomain = {
...
+    status: string = ServiceInstanceCreationStatus.Ready
   ): Promise<ServiceInstanceId> => {
     const id = uuidv4() as ServiceInstanceId;
     await db(context, 'ServiceInstance').insert([
       {
         id,
         name: serviceInstanceNameMappedByPlatformIdentifier[platformIdentifier],
         description: '',
-        creation_status: 'READY',
+        creation_status: status,
```
Comment (Kroustille) [inline apps/portal-api/src/modules/services/instances/domain.ts]: If you apply the suggestion with the function parameters:
```suggestion
        creation_status,
```

File: apps/portal-api/src/modules/services/deployments/deployments.domain.ts [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  DeploymentRequestMutator,
+} from '../../../model/kanel/public/DeploymentRequest';
+
+export const insertDeploymentRequest = async (
+  deploymentRequest: DeploymentRequestInitializer
+) => {
+  return dbUnsecure<DeploymentRequest>('DeploymentRequest')
+    .insert(deploymentRequest)
+    .returning('*');
+};
```
Comment (Kroustille) [inline apps/portal-api/src/modules/services/deployments/deployments.domain.ts]: This way you don't have to destructure the returned array in the caller. 
I personnally find it cleaner but it's up to you if you want to do the change 😉 
```suggestion
export const insertDeploymentRequest = async (
  deploymentRequest: DeploymentRequestInitializer
) => {
  const [createdDeploymentRequest] = await dbUnsecure<DeploymentRequest>('DeploymentRequest')
    .insert(deploymentRequest)
    .returning('*');
  return createdDeploymentRequest
};
```

File: apps/portal-api/knexfile.ts (hunk 3 above)
Comment (carinelebas) [reply apps/portal-api/knexfile.ts]: This code is currently only used for register and unregister (and now deployment), which are the only places transaction is set in the context. 
Without the check `trx.isCompleted()`, it will throw an error if we try to execute queries on a completed transaction (I had the issues in tests). With it, it should be safe, since query will be executed outside the transaction.
We need a better way to manage the transactions, though to better manage the closed transaction. But I'd rather tackle it in tech pains. 
Wdyt?

File: apps/portal-api/src/modules/services/deployments/deployments.app.ts (hunk 2 above)
Comment (carinelebas) [reply apps/portal-api/src/modules/services/deployments/deployments.app.ts]: They are not the same objects, and don't have the same fields: one is the kanel model (with fields to support platform data), the other one is the graphql object (only what is needed by the front). I would like to have an easier solution, but like this the conversion is clear at least.

File: apps/portal-api/src/modules/services/instances/domain.ts (hunk 4 above)
Comment (carinelebas) [reply apps/portal-api/src/modules/services/instances/domain.ts]: done in [f19ea09](https://github.com/FiligranHQ/xtm-hub/pull/1102/commits/f19ea09d05939af65f1ad757710dbc51dca16a14)

File: apps/portal-api/knexfile.ts (hunk 3 above)
Comment (Kroustille) [reply apps/portal-api/knexfile.ts]: Yes I agree it's better to tackle it in tech pain and it can be a broader subject 👍

**Review by hervyt — APPROVED** [review]
